SCF/SCF-Session6/BonusProblem1-Mutable/MutablePoly.java

```diff
@@ -0,0 +1,197 @@
+		private int coef;
+		private int power;
+		Term next;
+		
+		Term(int coef, int power){
+			this.coef = coef;
+			this.power = power;
+			this.next = null;
+		}
+	}
+	
+	/*
+	 * Add Term in the Polynomial list
+	 * @param  coef- Coefficient of term
+	 * @param  power - power of term
+	 * @return nothing
+	 */
+	public void addTerm(int coef, int power) {
+		Term term = new Term(coef, power);
+		if(firstTerm == null) {
+			firstTerm = term;
+			lastTerm = term;
+		}
+		else {
+			Term preTerm = this.firstTerm;
+			while(preTerm != null) {
+				if(preTerm.power == power) {
+					preTerm.coef += coef;
+					break;
+				}
+				preTerm = preTerm.next;
+			}
+			if(preTerm == null) {
+				lastTerm.next = term;
+				lastTerm = lastTerm.next;
+			}
+		}
+	}
+	
+	/*
+	 * Delete Term from Polynomial list
+	 * @param  coef- Coefficient of term
+	 * @param  power - power of term
+	 * @return Term - Deleted term reference  
+	 */
+	public boolean removeTerm(int coef, int power) {
+		Term term = this.firstTerm;
+		Term preTerm = this.firstTerm;
+		if(this.firstTerm.coef == coef && this.firstTerm.power == power) {
+			this.firstTerm = this.firstTerm.next;
+			term.next = null;
+			term = null;
+			return true;
+		}
+		while(term != null) {
+			if(term.coef == coef && term.power == power) {
+				preTerm.next = term.next;
+				if(term == this.lastTerm) {
+					this.lastTerm = preTerm;
+				}
+				term.next = null;
+				term = null;
+				return true;
+			}
+			preTerm = term;
+			term = term.next;
+		}
+		return false;
+	}
+	
+	/*
+	 * Print All Term of polynomial list
+	 * @param  nothing
+	 * @return nothing
+	 */
+	public void printPolynomial() {
+		Term term = this.firstTerm;
+		while(term != null) {
+			String sign = (term != this.firstTerm ? (term.coef >0 ? " + ":" - ") :(term.coef >0 ? "":" - "));
+			System.out.print(sign + Math.abs(term.coef)+ "*X^" + term.power + " ");
+			term = term.next;
+		}
+		System.out.println("");
+	}
+	
+	/*
+	 * Computer value of Polynomial for given variable value
+	 * @param  valueOfVariable - value of variable 
+	 * @return double - total polynomial value at variable value
+	 */
+	public double evaluate(double valueOfVariable) {
+		double totalValue = 0.0;
+		Term term = this.firstTerm;
+		while(term != null) {
+			totalValue += (double)term.coef * Math.pow((double)valueOfVariable, (double)term.power);
+			term = term.next;
+		}
+		return totalValue;
+	}
+	
+	/*
+	 * Compute Degree of Polynomial
+	 * @param  nothing
+	 * @return int - Degree of Polynomial
+	 */
+	public int degree() {
+		Term term = this.firstTerm;
+		int maxDegree = -1;
+		while(term != null) {
+			maxDegree = Math.max(maxDegree, term.power);
+			term = term.next;
+		}
+		return maxDegree;
+	}
+	
+	/*
+	 * Addition of two Mutable Polynomial
+	 * @param poly2 - Second Polynomial list
+	 * @return MutablePoly - Result of Addition of Two Polynomial
+	 */
+	public MutablePoly addPoly(MutablePoly poly2) {
+		MutablePoly poly3 = new MutablePoly();
```

Comment:
you are adding new term at the end everytime, new term should be added taking its power in account. I mean if I'm adding x^3 to x^2 + 1 this will result x^2 + 1 + x^3 as per your logic.

Reply:
For now its fine, no need to implement, as it is not impacting functionality or calculations.